packages/clerk-js/src/utils/authPropHelpers.ts

```diff
@@ -1,69 +1,36 @@
-import { camelToSnake } from '@clerk/shared';
-import type { ClerkOptions, DisplayConfigResource } from '@clerk/types';
-import type { ParsedQs } from 'qs';
+import type { DisplayConfigResource } from '@clerk/types';
 import qs from 'qs';
 
-import { hasBannedProtocol, isAllowedRedirectOrigin, isValidUrl } from './url';
+import { hasBannedProtocol, isValidUrl } from './url';
 
-type PickRedirectionUrlKey = 'afterSignUpUrl' | 'afterSignInUrl' | 'signInUrl' | 'signUpUrl';
-
-type PickRedirectionOptions = {
-  queryParams?: ParsedQs;
-  displayConfig?: DisplayConfigResource;
-  options?: ClerkOptions;
-  ctx?: any;
+type PickUrlOptions = {
+  validator?: (url: string) => boolean;
+  formatter?: (url: string) => string;
 };
 
-/**
- *
- * pickRedirectionProp(key, options, accessRedirectUrl)
- *
- * Retrieve auth redirect props passed through querystring parameters, component props, ClerkProvider props
- * or display config settings.
- *
- * The priority is:
- * 1. Querystring parameters
- * 2. ClerkJS component props
- * 3. ClerkProvider props
- * 4. Display configuration payload
- */
-export const pickRedirectionProp = (
-  key: PickRedirectionUrlKey,
-  { ctx, queryParams, displayConfig, options }: PickRedirectionOptions,
-  accessRedirectUrl = true,
-): string => {
-  const snakeCaseField = camelToSnake(key);
-  const queryParamValue = queryParams?.[snakeCaseField];
-
-  const primaryQueryParamRedirectUrl = typeof queryParamValue === 'string' ? queryParamValue : undefined;
-  const secondaryQueryParamRedirectUrl =
-    accessRedirectUrl && typeof queryParams?.redirect_url === 'string' ? queryParams.redirect_url : undefined;
-
-  let queryParamUrl: string | undefined;
-  if (
-    primaryQueryParamRedirectUrl &&
-    isAllowedRedirectOrigin(primaryQueryParamRedirectUrl, options?.allowedRedirectOrigins)
-  ) {
-    queryParamUrl = primaryQueryParamRedirectUrl;
-  } else if (
-    secondaryQueryParamRedirectUrl &&
-    isAllowedRedirectOrigin(secondaryQueryParamRedirectUrl, options?.allowedRedirectOrigins)
-  ) {
-    queryParamUrl = secondaryQueryParamRedirectUrl;
-  }
-
-  const url =
-    queryParamUrl ||
-    ctx?.[key] ||
-    (accessRedirectUrl ? ctx?.redirectUrl : undefined) ||
-    options?.[key] ||
-    displayConfig?.[key];
-
-  if (!isValidUrl(url, { includeRelativeUrls: true }) || hasBannedProtocol(url)) {
-    return '';
-  }
-
-  return url;
+export const pickUrl = (key: string | string[], source: Record<string, any>, opts?: PickUrlOptions): string => {
```

Comment:
🔧  Options for `pickUrl` util could be an object like:

```ts
type PickUrlParams = {
  key: string | string[];
  source: Record<string, any>;
  opts?: PickUrlOptions
}
```

Reply:
I don't think the two first params can change in the future, so I just think this will make it so we have to type `key` and `source` every time.